File: src/Python/101-200/163.MissingRanges.py

class Solution:
    def findMissingRanges(self, nums, lower, upper):
        expected = lower
        newUpper = upper
        ret = []
        for i in range(len(nums) + 1):
            if i < len(nums) and nums[i] < expected:
                continue
            val = newUpper + 1 if i == len(nums) else nums[i]
            if expected != val and expected + 1 == val:
                ret.append(str(expected))
            elif expected != val:
                ret.append(str(expected) + '->' + str(val - 1))
            expected = val + 1
        return ret


if __name__ == "__main__":
    solution = Solution()
    print(solution.findMissingRanges([0, 1, 3, 50, 75], 0, 99))

File: src/Java/101-200/163.MissingRanges.java

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;

public class Solution {
    List<String> findMissingRanges(int[] nums, int lower, int upper) {
        long expected = lower, newUpper = upper;
        List<String> ret = new LinkedList<>();
        for (int i = 0; i <= nums.length; i++) {
            if (i < nums.length && nums[i] < expected)
                continue;
            long val = (i == nums.length) ? newUpper + 1 : nums[i];
            if (expected != val && expected + 1 == val) {
                ret.add(String.valueOf(expected));
            } else if (expected != val) {
                ret.add(String.valueOf(expected) + "->" + String.valueOf(val - 1));
            }
            expected = val + 1;
        }
        return ret;
    }
}
